syntax = "proto3";

package services.hogwords;
option go_package = "github.com/verloop/api/go/services/hogwords";

service SelectFromChoice {
	rpc GetChoicesMatch (Query) returns (Response);
}

message ListOfStrings {
	repeated string choices=1;
}

message Choice {
	string text = 1;
	double score = 2;
}

message ListOfChoices {
	repeated Choice choices = 1;
	float recommendedThreshold = 2;
}

message Response {

	string msgId = 1; // Response msg Id, should be same as query msg id
	string rid = 2; 	// Message chat room id: rid
	string clientId = 3; 	// _tenantId, e.g. nykaa or decathlonblre 
	Choice selectedChoice = 4;
	bool matchFound = 5; // if none of the strings clear threshold, set to False
	ListOfChoices choiceScoreList = 6;
	
	InputTypes inputType = 7; 	// default is GENERIC
	ReturnModes returnMode = 9;	// default is EXTRACTONE
	
	//TODO: add field for created timestamp
}

enum InputTypes {
	/* 
	* Note that enum values use C++ scoping rules i.e.
	* UNKNOWN from ReturnModes conflicts with new definition here
	* But each enum must begin zero value, and hence the specific names for UNKNOWNS :)
	*/

	INPUTTYPE_UNKNOWN = 0;

	GENERIC = 1;
	
	YESNO = 2; // "Yeah, yes, sure, ok, :thumbsup emoji:", "No, nope, nah"
	
	RATING3 = 3; // "Bad", "Okay", "Great"
	
	// 4 enum value is empty
	
	RATING5 = 5; // "Not satisfied", "Somewhat sat", "Ok", "Sat", "Amazing"
}

enum ReturnModes {
	/* 
	* Note that enum values use C++ scoping rules i.e.
	* UNKNOWN from InputTypes conflicts with new definition here
	* But each enum must begin zero value, and hence the specific names for UNKNOWNS :)
	*/

	RETURNMODE_UNKNOWN = 0;
	EXTRACTONE = 1;
	SCOREALL = 2;
}

message Query {
	// Query id
	string msgId = 1;

	// Message chat room id: rid
	string rid = 2;
	
	// _tenantId, e.g. nykaa or decathlonblre 
	string clientId = 3;

	// input to match against
	string queryString = 4;

	ListOfStrings choiceStrings = 5;
	
	bool ignoreNumbers = 6;

	// default is GENERIC
	InputTypes inputType = 7;

	// default is EXTRACTONE
	ReturnModes returnMode = 9;
}